File: core/src/com/infeez/simple/Cells.java
package com.infeez.simple;

import com.infeez.simple.base.GameSpriteBatch;
import com.infeez.simple.entity.Cell;
import com.infeez.simple.utils.BoardArrayPosition;
import com.infeez.simple.utils.Constants.GameEnvTypes;

import javax.annotation.Nonnull;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

public class Cells implements Iterable<Cell> {

    private static final int COUNT_CELL = 8;
    private final Cell[][] cells = new Cell[COUNT_CELL][COUNT_CELL];

    public Cells() {
    }

    public void createBoard(@Nonnull GameSpriteBatch batch) {
        float x = 0, y = 0;
        for (int i = 0; i < COUNT_CELL; i++) {
            for (int j = 0; j < COUNT_CELL; j++) {
                GameEnvTypes type = GameEnvTypes.BLACK;
                if ((j % 2 == 1 && i % 2 == 1) || (j % 2 == 0 && i % 2 == 0)) {
                    type = GameEnvTypes.WHITE;
                }
                Cell cell = new Cell(i, j, x, y, type, batch);
                setCell(cell, i, j);
                x += Cell.CELL_SIZE;
            }
            x = 0;
            y += Cell.CELL_SIZE;
        }
    }

    public void startCellsPosition() {
        clearCheckers();
        for (int i = 0; i < COUNT_CELL; i++) {
            for (int j = 0; j < COUNT_CELL; j++) {
                if (i <= 2 && cells[i][j].isBlackType()) {
                    cells[i][j].setChecker(GameEnvTypes.BLACK);
                } else if (i >= 5 && cells[i][j].isBlackType()) {
                    cells[i][j].setChecker(GameEnvTypes.WHITE);
                }
            }
        }
    }

    public Cell getCell(@Nonnull BoardArrayPosition boardArrayPosition) {
        return cells[boardArrayPosition.getIndexFirst()][boardArrayPosition.getIndexSecond()];
    }

    public void setCell(@Nonnull Cell cell, int i, int j) {
        cells[i][j] = cell;
    }

    public void clearCheckers() {
        for (Cell cell : this) {
            cell.removeChecker();
        }
    }

    public Cell findCellByCoordinatesAndHaveChecker(int x, int y) {
        for (Cell cell : this) {
            if (cell.contains(x, y) && cell.isChecker()) {
                return cell;
            }
        }
        return null;
    }

    public List<Cell> toList() {
        List<Cell> result = new ArrayList<>(COUNT_CELL + COUNT_CELL);
        for (Cell[] a : cells) {
            Collections.addAll(result, a);
        }
        return result;
    }

    public Iterator<Cell> iterator() {
        return toList().iterator();
    }
}

File: core/src/main/kotlin/com/infeez/simple/Cells.kt
package com.infeez.simple

import com.infeez.simple.base.GameSpriteBatch
import com.infeez.simple.entity.Cell
import com.infeez.simple.utils.BoardArrayPosition
import com.infeez.simple.utils.BoardConfig
import com.infeez.simple.utils.Constants.GameEnvTypes

class Cells : Iterable<Cell> {
    private val cells: Array<Array<Cell?>> = Array(BoardConfig.BOARD_SIZE) {
        arrayOfNulls(BoardConfig.BOARD_SIZE)
    }

    fun createBoard(batch: GameSpriteBatch? = null) {
        var y = 0f
        for (row in 0 until BoardConfig.BOARD_SIZE) {
            var x = 0f
            for (column in 0 until BoardConfig.BOARD_SIZE) {
                val type = if ((column + row) % 2 == 0) {
                    GameEnvTypes.WHITE
                } else {
                    GameEnvTypes.BLACK
                }
                val cell = Cell(column, row, x, y, type, batch)
                setCell(cell, column, row)
                x += BoardConfig.CELL_SIZE
            }
            y += BoardConfig.CELL_SIZE
        }
    }

    fun startCellsPosition() {
        clearCheckers()
        for (cell in this) {
            val row = cell.boardPosition.indexSecond
            if (row <= 2 && cell.isBlackType()) {
                cell.setChecker(GameEnvTypes.BLACK)
            } else if (row >= 5 && cell.isBlackType()) {
                cell.setChecker(GameEnvTypes.WHITE)
            }
        }
    }

    fun getCell(boardArrayPosition: BoardArrayPosition): Cell {
        require(boardArrayPosition.indexFirst in 0 until BoardConfig.BOARD_SIZE) {
            "Board file index must be in range 0..7."
        }
        require(boardArrayPosition.indexSecond in 0 until BoardConfig.BOARD_SIZE) {
            "Board rank index must be in range 0..7."
        }
        return cells[boardArrayPosition.indexFirst][boardArrayPosition.indexSecond]
            ?: error("Board has not been created.")
    }

    fun setCell(cell: Cell, i: Int, j: Int) {
        cells[i][j] = cell
    }

    fun clearCheckers(type: GameEnvTypes? = null) {
        for (cell in this) {
            if (type == null || cell.checker?.type == type) {
                cell.removeChecker()
            }
        }
    }

    fun findCellByCoordinates(x: Float, y: Float): Cell? {
        return firstOrNull { cell ->
            cell.contains(x, y)
        }
    }

    fun findCellByCoordinatesAndHaveChecker(x: Float, y: Float): Cell? {
        return firstOrNull { cell ->
            cell.contains(x, y) && cell.isChecker()
        }
    }

    fun toList(): List<Cell> {
        val result = ArrayList<Cell>(BoardConfig.BOARD_SIZE * BoardConfig.BOARD_SIZE)
        for (column in cells) {
            for (cell in column) {
                if (cell != null) {
                    result.add(cell)
                }
            }
        }
        return result
    }

    override fun iterator(): Iterator<Cell> = toList().iterator()
}
